refactor(sticky-roles): centralize all logic and fix everything

kyranet opened the pull request.

kyranet committed:
refactor(sticky-roles): centralize all logic and fix everything

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+		if (arrayIndex === -1) return [];
+
+		const entry = this.entries[arrayIndex];
+		await this.#guild.settings.update(GuildSettings.StickyRoles, entry, { arrayAction: 'remove', arrayIndex, extraContext });
```
![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif) ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+		if (entry.roles.length === roles.length) return entry.roles;
+
+		const clone: StickyRole = { user: userID, roles };
+		await this.#guild.settings.update(GuildSettings.StickyRoles, clone, { arrayIndex, extraContext });
```
For consistency please wrap this over multiple lines like in other places where update is used

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+				user: userID,
+				roles: [roleID]
+			};
+			await this.#guild.settings.update(GuildSettings.StickyRoles, entry, { arrayAction: 'add', extraContext });
```
![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

favna replied:
`arrayAction: ArrayActions.Add` use the enum from klasa

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+			user: userID,
+			roles: [...this.addRole(roleID, entry.roles)]
+		};
+		await this.#guild.settings.update(GuildSettings.StickyRoles, clone, { arrayIndex, extraContext });
```
![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+
+		// Remove entry if empty, update otherwise
+		if (clone.roles.length === 0) {
+			await this.#guild.settings.update(GuildSettings.StickyRoles, entry, { arrayAction: 'remove', arrayIndex, extraContext });
```
![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif) ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+		if (clone.roles.length === 0) {
+			await this.#guild.settings.update(GuildSettings.StickyRoles, entry, { arrayAction: 'remove', arrayIndex, extraContext });
+		} else {
+			await this.#guild.settings.update(GuildSettings.StickyRoles, clone, { arrayIndex, extraContext });
```
![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

kyranet committed:
refactor(serializers): fix stickrole's Serializer#validate

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+	public async add(userID: string, roleID: string, extraContext?: unknown): Promise<readonly string[]> {
+		const arrayIndex = this.getKey(userID);
+		if (arrayIndex === -1) {
+			const entry: StickyRole = {
```
To avoid confusion of other variables also called entry in this file with wholly different context I suggest renaming this variable

favna commented on src/lib/structures/managers/StickyRoleManager.ts:
```diff
@@ -0,0 +1,117 @@
+		return this.entries.find(entry => entry.user === userID)?.roles ?? [];
+	}
+
+	public async fetch(userID: string, extraContext?: unknown): Promise<readonly string[]> {
```
extraContent should be typed like this because we're always passing in the author (and if we're not then we should)

```ts
type ExtraContent<T> = T & { author: string }
```

kyranet committed:
refactor(sticky-roles): resolved requested changes

kyranet committed:
fix(sticky-roles): fix last edge case
Essentially what would happen if another user had an invalid role, if we patch with Overwrite, all entries will be checked.

favna commented on src/serializers/stickyrole.ts:
```diff
@@ -10,14 +10,7 @@ export default class extends Serializer {
-			return data;
-
-		}
+			&& data.roles.every(role => typeof role === 'string' && guild!.roles.has(role))) return data;
```
What will happen if the user has a sticky role that has been removed now that you revert this...

kyranet replied:
It's covered by the `StickyRoleManager`, if you read the comments, it's theoretically impossible for that to ever return false.

This check is there so people don't push a StickyRole entry with +1000 invalid roles via the dashboard's API.

kyranet committed:
refactor(sticky-roles): optimize code
